src/azure-cli-core/azure/cli/core/azclierror.py

```diff
@@ -4,85 +4,178 @@
         self.error_msg = error_msg
-        self.raw_exception = raw_exception
-        self.command = command
+        # set recommendations to fix the error if the message is not actionable,
+        # they will be printed to users after the error message, one recommendation per line
         self.recommendations = []
+        # exception trace for the error
+        self.exception_trace = None
         super().__init__(error_msg)
 
     def set_recommendation(self, recommendation):
         self.recommendations.append(recommendation)
 
-    def set_raw_exception(self, raw_exception):
-        self.raw_exception = raw_exception
+    def set_exception_trace(self, exception_trace):
+        self.exception_trace = exception_trace
 
     def print_error(self):
         from azure.cli.core.azlogging import CommandLoggerContext
         with CommandLoggerContext(logger):
-            message = '{}: {}'.format(self.error_type.value, self.error_msg)
+            # print error type and error message
+            message = '{}: {}'.format(self.__class__.__name__, self.error_msg)
             logger.error(message)
-            if self.raw_exception:
-                logger.exception(self.raw_exception)
+            # print exception trace if there is
+            if self.exception_trace:
+                logger.exception(self.exception_trace)
+            # print recommendations to action
             if self.recommendations:
                 for recommendation in self.recommendations:
                     print(recommendation, file=sys.stderr)
 
     def send_telemetry(self):
-        import azure.cli.core.telemetry as telemetry
-        telemetry.set_error_type(self.error_type.value)
-
-        # For userfaults
-        if self.error_type in [AzCLIErrorType.CommandNotFoundError,
-                               AzCLIErrorType.ArgumentParseError,
-                               AzCLIErrorType.ValidationError,
-                               AzCLIErrorType.ManualInterrupt]:
-            telemetry.set_user_fault(self.error_msg)
-
-        # For failures: service side error, client side error, unexpected error
-        else:
-            telemetry.set_failure(self.error_msg)
-
-        # For unexpected error
-        if self.raw_exception:
-            telemetry.set_exception(self.raw_exception, '')
+        telemetry.set_error_type(self.__class__.__name__)
+# endregion
+
+
+# region: Second Layer
+# Main categories of the AzureCLI error types, used for Telemetry alalysis
+# DO NOT raise the error classes here directly in your codes.
+class UserFault(AzCLIError):
+    """ Users should be responsible for the errors. """
+    def send_telemetry(self):
+        super().send_telemetry()
+        telemetry.set_user_fault(self.error_msg)
+
+
+class ServiceError(AzCLIError):
+    """ Azure Services should be responsible for the errors.  """
+    def send_telemetry(self):
+        super().send_telemetry()
+        telemetry.set_failure(self.error_msg)
```

Comment:
so in telemetry, service error and client error will be failure?

Reply:
when service error, no need to set exception trace?

Reply:
> so in telemetry, service error and client error will be failure?

yes

Reply:
> when service error, no need to set exception trace?

yes, exception trace is only set when the error is not caught. If an error was caught, that means we know what happens there,  and I think there is no need to set exception trace any more.

Reply:
is there easy way to distinguish them, such as one field? otherwise we'll need to maintain a list of error types which are client, which are service.

Reply:
sure, we can use set_client_error() and set_service_error() to replace set_failure, which should make the result be one of ['Success', 'UserFault', 'ClientError', 'ServiceError'].